internal/helpers/secrets.go

```diff
@@ -169,11 +262,38 @@ func SyncSecret(ctx context.Context, client ctrlclient.Client, obj ctrlclient.Ob
 
 	if exists {
 		logger.V(consts.LogLevelDebug).Info("Updating secret")
-		return client.Update(ctx, &dest)
+		if err := client.Update(ctx, &dest); err != nil {
```

Comment:
One thing that is really powerful about the Kubernetes API's is our ability to use Label filtering and push that computation out to the kube api server.

In `FindSecretsOwnedByObj` we're doing a `client.List()` with `OwnerLabels` but that will still return every Kubernetes secret (in this namespace) that is managed by VSO.
If we add an extra "Owner label" on line 252 that directly references the VSO Secret we can dramatically reduce the work needed in `FindSecretsOwnedByObj` and have `client.List()` return *only* the Kube Secrets that we want to interact with. I imagine that there could be thousands of Kube secrets managed in a single ns and we'd be iterating through that list of 1000's of secrets every time this code triggers otherwise. (Assuming I'm reading the code correctly?)

 
I'm thinking something like this (again Line 252):
```
	// always add the "owner" labels last to guard against intersections with meta.Destination.Labels
	for k, v := range OwnerLabels {
		labels[k] = v
	}
        labels["secrets.hashicorp.com/ownerRef"] = obj.GetName()
```
..and then add it to the OwnerLables again in `FindSecretsOwnedByObj` when you're about to issue the `client.List()`?

Thoughts?

Reply:
> One thing that is really powerful about the Kubernetes API's is our ability to use Label filtering and push that computation out to the kube api server.

Indeed. I think we are already benefiting from that feature with our use of `OwnerLabels`

> 
> In `FindSecretsOwnedByObj` we're doing a `client.List()` with `OwnerLabels` but that will still return every Kubernetes secret (in this namespace) that is managed by VSO. If we add an extra "Owner label" on line 252 that directly references the VSO Secret we can dramatically reduce the work needed in `FindSecretsOwnedByObj` and have `client.List()` return _only_ the Kube Secrets that we want to interact with. I imagine that there could be thousands of Kube secrets managed in a single ns and we'd be iterating through that list of 1000's of secrets every time this code triggers otherwise. (Assuming I'm reading the code correctly?)
> 
> I'm thinking something like this (again Line 252):
> 
> ```
> 	// always add the "owner" labels last to guard against intersections with meta.Destination.Labels
> 	for k, v := range OwnerLabels {
> 		labels[k] = v
> 	}
>         labels["secrets.hashicorp.com/ownerRef"] = obj.GetName()
> ```
> 
> ..and then add it to the OwnerLables again in `FindSecretsOwnedByObj` when you're about to issue the `client.List()`?
> 
> Thoughts?

Sounds like a pretty nice optimization, although that means we would miss the secrets that do not have that label. Presumably the next time they are sync'd it would be added.

One issue I see with the proposal to use the object's name is that we would end up with orphans whenever the CR's name was updated. We could probably mitigate that issue if we use the object's UID instead of its Name.

Reply:
added in https://github.com/hashicorp/vault-secrets-operator/pull/234/commits/fcc25d8cb3765c9d5e1e9165d5f724d79e16caea